From 9572dd658ceb88ea7a151e7a1dbc2ab770a4f4d1 Mon Sep 17 00:00:00 2001
From: Erico Nunes <nunes.erico@gmail.com>
Date: Thu, 19 Dec 2019 22:51:07 +0100
Subject: [PATCH 036/102] lima: set shader caps to optimize control flow

With these new caps, nir is able to unroll loops and optimize
conditionals much more efficiently in both gpit and ppir.
panfrost and vc4 were used as reference for the values.

Signed-off-by: Erico Nunes <nunes.erico@gmail.com>
Reviewed-by: Vasily Khoruzhick <anarsoul@gmail.com>
Tested-by: Marge Bot <https://gitlab.freedesktop.org/mesa/mesa/merge_requests/3176>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/merge_requests/3176>
---
 src/gallium/drivers/lima/lima_screen.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/gallium/drivers/lima/lima_screen.c b/src/gallium/drivers/lima/lima_screen.c
index 4dc1fc652b8..9b6335ee24c 100644
--- a/src/gallium/drivers/lima/lima_screen.c
+++ b/src/gallium/drivers/lima/lima_screen.c
@@ -178,6 +178,9 @@ get_vertex_shader_param(struct lima_screen *screen,
    case PIPE_SHADER_CAP_MAX_TEX_INDIRECTIONS:
       return 16384; /* need investigate */
 
+   case PIPE_SHADER_CAP_MAX_CONTROL_FLOW_DEPTH:
+      return 1024;
+
    case PIPE_SHADER_CAP_MAX_INPUTS:
       return 16; /* attributes */
 
@@ -186,6 +189,7 @@ get_vertex_shader_param(struct lima_screen *screen,
 
    case PIPE_SHADER_CAP_MAX_CONST_BUFFER_SIZE:
       return 4096; /* need investigate */
+
    case PIPE_SHADER_CAP_MAX_CONST_BUFFERS:
       return 1;
 
@@ -195,6 +199,9 @@ get_vertex_shader_param(struct lima_screen *screen,
    case PIPE_SHADER_CAP_MAX_TEMPS:
       return 256; /* need investigate */
 
+   case PIPE_SHADER_CAP_MAX_UNROLL_ITERATIONS_HINT:
+      return 32;
+
    default:
       return 0;
    }
@@ -214,8 +221,12 @@ get_fragment_shader_param(struct lima_screen *screen,
    case PIPE_SHADER_CAP_MAX_INPUTS:
       return LIMA_MAX_VARYING_NUM - 1; /* varying, minus gl_Position */
 
+   case PIPE_SHADER_CAP_MAX_CONTROL_FLOW_DEPTH:
+      return 1024;
+
    case PIPE_SHADER_CAP_MAX_CONST_BUFFER_SIZE:
       return 4096; /* need investigate */
+
    case PIPE_SHADER_CAP_MAX_CONST_BUFFERS:
       return 1;
 
@@ -236,6 +247,9 @@ get_fragment_shader_param(struct lima_screen *screen,
    case PIPE_SHADER_CAP_INDIRECT_OUTPUT_ADDR:
       return 0;
 
+   case PIPE_SHADER_CAP_MAX_UNROLL_ITERATIONS_HINT:
+      return 32;
+
    default:
       return 0;
    }
-- 
2.24.1

